### DISCUSSION TITLE
How to manage dependencies between plugins

### DISCUSSION THREAD

**Discussion by peanubnutter** [runelite/runelite#16706] [Q&A]
Hi, I'm developing a new Collection Log related plugin, but I want it to be compatible with the existing Collection Log plugin.

It seems like I can at least detect whether Collection Log is installed by checking [ExternalPluginManager](https://static.runelite.net/api/runelite-client/net/runelite/client/externalplugins/ExternalPluginManager.html) on startUp() and subscribing to [onPluginChanged](http://static.runelite.net/api/runelite-client/net/runelite/client/events/class-use/PluginChanged.html) to check if it becomes loaded or unloaded (though this is not very well documented, except through its use in VirtualLevelsPlugin). 

I have a few questions:

1.) Is the above method reasonable, or are there any corner cases to watch out for?

2.) Can (or should I) use [@PluginDependency](http://static.runelite.net/api/runelite-client/net/runelite/client/plugins/PluginDependency.html) to depend on CollectionLogPlugin.class? It's not an "official" subclass of [Plugin](https://static.runelite.net/api/runelite-client/net/runelite/client/plugins/Plugin.html), so is this still possible?

2.a) What does a "dependency" mean, exactly? Would it be possible to install my plugin at all, or would RuneLite prompt the user to also install dependencies, or would they also be installed automatically? If I enable my plugin, does it auto-enable any dependent plugins, etc?

2.b) Would I need to import the plugin code in some way, like by modifying build.gradle? I haven't seen any examples of this before.

3.info) CollectionLogPlugin uses [onScriptPostFired](https://static.runelite.net/api/runelite-api/net/runelite/api/events/ScriptPostFired.html) with ScriptId.COLLECTION_DRAW_LIST (with [clientThread.invokeLater](https://static.runelite.net/api/runelite-client/net/runelite/client/callback/ClientThread.html#invokeLater(java.lang.Runnable))) to redraw elements of the collection log.

3.a) I haven't really seen an explanation for why invokeLater is necessary here, rather than just directly running the code. Does the code need to be run on the game thread, otherwise the collection log stuff isn't available to be modified?

3.b) I also want to call this function to further modify the collection log display widget. How do I guarantee that my plugin's code is called *after* CollectionLogPlugin (if that plugin is also enabled), if they both call invokeLAter inside onScriptPostFired?

4.) Is there a Discord or some community to ask for help besides here? I've been able to figure most things out so far.

**Comment by geheur** [comment]
runelite.net/discord

**Reply by geheur** [reply]
you can't use @plugindependency because you can't reference classes from other hub plugins.
3b probably use a negative priority onscriptpostfired subscriber. I assume invokelaters are handled in the order they're put in.
dunno about the rest. inter-hub-plugin stuff isn't well supported.

**Reply by peanubnutter** [reply]
Thanks for the quick response and the Discord link! I didn't know I could set priority there, thanks.

**Reply by peanubnutter** [reply]
For anyone finding this on Google later: negative priority onScriptPostFired did work successfully. The default priority is 0, so my plugin will always run second now.

**Comment by Adam-** [comment]
1) I think the best way is to use the plugin descriptor `conflicts` with the name of the other plugin.
2) You cannot access other pluginhub plugin classes because they are in a separate classloader
3) a. Maybe it isn't, you would have to ask them. Or try it.
3) b. Set the priority on the subscriber lower/higher than theirs
4) https://runelite.net/discord

**Reply by peanubnutter** [reply]
Thank you!

1) I do want to be able to use both plugins simultaneously, so I didn't want to use conflicts. It could be a last resort if I can't figure out how to make them compatible.

I'll try a couple of the other solutions, thank you!

**Reply by Nightfirecat** [reply]
You can read configs via ConfigManager (and therefore react to ConfigChanged events if one plugin's config is updated), or theoretically post events with object types which aren't defined in the plugins. (eg. you can likely post a `String` to the event bus and pick it up from another plugin)
